### PR TITLE
New action: actions/haskell-setup

### PR DESCRIPTION
**Pull request by JackKelly-Bellroy**
Adds a new action `actions/haskell-setup`, mapping to the github [`actions/setup-haskell`](https://github.com/actions/setup-haskell). The name is swapped around for consistency with `actions/java-setup` in this repo, but it might be a bad idea to establish a pattern where the `github-actions-dhall` names differ from the actual names of actions. I'm willing to rename this to `setup-haskell` and also do `java-setup` -> `setup-java` if you think that's better.

I've also needed to import more things from prelude here. Let me know if you're happy with how I'm doing this, or if you'd rather some other way of getting the prelude (such as `/deps/Prelude.dhall` containing a single frozen import or something).

This PR is rebased atop #19, to hopefully avoid merge conflicts on the dhall SHA256s. Merge that first and the diff here won't have the cache noise. If you merge something else in-between, let me know and I will re-freeze and rebase this PR.

### CODE REVIEW

**Review by regadas — APPROVED** [review]
Thanks!

**Comment by regadas** [comment]
>  it might be a bad idea to establish a pattern where the github-actions-dhall names differ from the actual names of actions

I agree can you swap it around and rebase. If you could make a new PR for the `java-setup` that would be awesome.

File: steps/actions/cache.dhall [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
+in  λ(args : { path : Text, key : Text, hashFiles : List Text }) →
+      let hashFilesArg =
+            Text/concatMapSep "," Text (λ(x : Text) → "'${x}'") args.hashFiles
+
+      in  Step::{
+          , id = None Text
+          , name = Some "${args.path} cache"
+          , uses = Some "actions/cache@v2"
+          , run = None Text
+          , with = Some
```
Comment (regadas) [inline steps/actions/cache.dhall:15]:
```suggestion
          , `with` = Some
```

Comment (JackKelly-Bellroy) [reply steps/actions/cache.dhall:15]: Dhall 1.34.0's formatter and its linter both try to remove the backticks. Do you want them forced in?

**Comment by JackKelly-Bellroy** [comment]
Rebased atop #21 which fixes the `setup-foo` vs `foo-setup` switcharound.

There are a couple of things I'd like to talk about before merge:

1. The use of `actions/HaskellSetup` and `actions/setup-haskell`: do we want to push them into an `actions` record before they get too big? Having `actions/` as a quasi-namespace feels a bit weird, and AFAIK we can't assemble record keys from `Text` or anything.
2. `setup-haskell` would be the first step function to have a schema'd type as its argument. Are you OK with this, or should it accept a record-of-optionals? I feel like there's a lot of value to providing the default, so long as you're happy with the convention it establishes.

File: steps/actions/cache.dhall (hunk 1 above)
Comment (regadas) [reply steps/actions/cache.dhall]: Hm, wasn't aware of that... but looking at the release notes I can't see that  ... `with` is still a reserved keyword so escaping should still be necessary 🤷

**Review by regadas — APPROVED** [review]

**Comment by JackKelly-Bellroy** [comment]
Thanks. Also, I assume that since you've merged this PR, we're happy to have argument schemata, if I need to use those again on larger actions?

**Pull request merged**